File: app/src/main/java/com/alfianyusufabdullah/chatyuk/network/AuthRequest.java
package com.alfianyusufabdullah.chatyuk.network;

import android.support.annotation.NonNull;
import android.util.Log;

import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.auth.AuthResult;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.ValueEventListener;

import com.alfianyusufabdullah.chatyuk.interfaces.onAuthRequestListener;
import com.alfianyusufabdullah.chatyuk.model.ModelUser;
import com.alfianyusufabdullah.chatyuk.MyApplication;
import com.alfianyusufabdullah.chatyuk.utils.Constan;
import com.alfianyusufabdullah.chatyuk.utils.PreferencesManager;

public class AuthRequest {

    private static String Username;
    private static String Email;
    private static String Password;

    private static DatabaseReference databaseReference;
    private static ValueEventListener eventListener;

    public static AuthRequest Field(String username, String email, String password) {
        Username = username;
        Email = email;
        Password = password;
        return new AuthRequest();
    }

    public static AuthRequest Field(String email, String password) {
        Email = email;
        Password = password;
        return new AuthRequest();
    }

    public void startRegister(final onAuthRequestListener listener) {

        FirebaseAuth firebaseAuth = FirebaseAuth.getInstance();
        firebaseAuth.createUserWithEmailAndPassword(Email, Password)
                .addOnCompleteListener(new OnCompleteListener<AuthResult>() {
                    @Override
                    public void onComplete(@NonNull Task<AuthResult> task) {

                        if (task.isSuccessful()) {

                            String UserID = task.getResult().getUser().getUid();

                            ModelUser user = new ModelUser();
                            user.setUsername(Username);
                            user.setEmail(Email);
                            user.setUserId(UserID);
                            user.setLoginTime(Constan.getTime());

                            PreferencesManager.initPreferences().setUserInfo(user);
                            MyApplication.getFirebaseDatabaseReferences("user").child(UserID).setValue(user);

                            listener.onAuthSuccess();
                        } else {
                            listener.onAuthFailed(task.getException().getMessage());
                            Log.d("RegisterRequest", "onComplete: " + task.getException().getMessage());
                        }

                    }
                });

    }

    public void startLogin(final onAuthRequestListener listener) {

        FirebaseAuth auth = FirebaseAuth.getInstance();
        auth.signInWithEmailAndPassword(Email, Password)
                .addOnCompleteListener(new OnCompleteListener<AuthResult>() {
                    @Override
                    public void onComplete(@NonNull Task<AuthResult> task) {

                        if (task.isSuccessful()) {

                            final String UserID = task.getResult().getUser().getUid();

                            databaseReference = MyApplication.getFirebaseDatabaseReferences("user").child(UserID);
                            databaseReference.addListenerForSingleValueEvent(eventListener(listener, UserID));

                        } else {
                            listener.onAuthFailed(task.getException().getMessage());
                        }

                    }
                });
    }

    private ValueEventListener eventListener(final onAuthRequestListener listener, final String UserID) {
        eventListener = new ValueEventListener() {
            @Override
            public void onDataChange(DataSnapshot dataSnapshot) {

                ModelUser user = dataSnapshot.getValue(ModelUser.class);

                if (user != null) {
                    user.setLoginTime(Constan.getTime());
                    MyApplication.getFirebaseDatabaseReferences("user").child(UserID).setValue(user);
                    PreferencesManager.initPreferences().setUserInfo(user);

                    listener.onAuthSuccess();
                }

            }

            @Override
            public void onCancelled(DatabaseError databaseError) {

            }
        };

        return eventListener;
    }

    public static void removeSignInListener() {
        if (eventListener != null) {
            databaseReference.removeEventListener(eventListener);
        }
    }
}

File: app/src/main/java/com/alfianyusufabdullah/chatyuk/network/AuthRequest.kt
package com.alfianyusufabdullah.chatyuk.network

import android.util.Log
import com.google.firebase.database.DataSnapshot
import com.google.firebase.database.DatabaseError
import com.google.firebase.database.DatabaseReference
import com.google.firebase.database.ValueEventListener

import com.alfianyusufabdullah.chatyuk.interfaces.AuthRequestListener
import com.alfianyusufabdullah.chatyuk.model.ModelUser
import com.alfianyusufabdullah.chatyuk.MyApplication
import com.alfianyusufabdullah.chatyuk.utils.Constant
import com.google.firebase.auth.FirebaseAuth

class AuthRequest {

    companion object {

        private lateinit var databaseReference: DatabaseReference
        private lateinit var valueEventListener: ValueEventListener

        private lateinit var username: String
        private lateinit var email: String
        private lateinit var password: String

        fun field(email: String, password: String, username: String = "username"): AuthRequest {
            this.username = username
            this.email = email
            this.password = password

            return AuthRequest()
        }

        fun removeSignInListener() {
            if (this::valueEventListener.isInitialized) databaseReference.removeEventListener(valueEventListener)
        }

        private fun evenListener(listener: AuthRequestListener, userId: String): ValueEventListener {
            valueEventListener = object : ValueEventListener {
                override fun onCancelled(database: DatabaseError) {
                    listener.onAuthFailed(database.message)
                }

                override fun onDataChange(snapshot: DataSnapshot) {
                    with(snapshot.value as? ModelUser) {
                        this?.let {
                            it.loginTime = Constant.time
                            MyApplication.getFirebaseDatabaseReferences("user").child(userId).setValue(it)
                            listener.onAuthSuccess(it)
                        }
                    }
                }
            }

            return valueEventListener
        }
    }

    fun doRegister(listener: AuthRequestListener) {
        Log.d("EMAIL", email )
        Log.d("PASSWORD", password)
        FirebaseAuth.getInstance().createUserWithEmailAndPassword(email, password)
                .addOnCompleteListener {

                    if (it.isSuccessful) {
                        val userId = it.result?.user?.uid

                        val user = ModelUser()
                        user.username = username
                        user.email = email
                        user.userId = userId
                        user.loginTime = Constant.time

                        MyApplication.getFirebaseDatabaseReferences("user")
                                .child(userId ?: "0")
                                .setValue(user)

                        listener.onAuthSuccess(user)
                    } else {
                        listener.onAuthFailed(it.exception?.localizedMessage ?: "error")
                        Log.d("RegisterRequest", "onFailed: " + it.exception?.localizedMessage)
                    }
                }
    }

    fun doLogin(listener: AuthRequestListener) {
        FirebaseAuth.getInstance().signInWithEmailAndPassword(email, password)
                .addOnCompleteListener {
                    if (it.isSuccessful) {
                        val userId = it.result?.user?.uid

                        databaseReference = MyApplication.getFirebaseDatabaseReferences("user")
                                .child(userId ?: "0")
                        databaseReference.addListenerForSingleValueEvent(
                                evenListener(listener, userId ?: "0")
                        )
                    }
                }
    }
}
